File: Recursion/101_SymmetricTree.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        return self.helper(root, root)

    # If two nodes are symmetric
    def helper(self, lNode, rNode):
        if not lNode or not rNode:
            return lNode == rNode
        if lNode.val != rNode.val:
            return False
        return (self.helper(lNode.left, rNode.right) and
                self.helper(lNode.right, rNode.left))

"""
[]
[1]
[1,2,3]
[1,2,2,3,4,4,3]
[1,2,2,null,3,null,3]
"""

File: Recursion/101_SymmetricTree.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-29 13:51:57
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return helper(root, root);
    }

    bool helper(TreeNode* lNode, TreeNode* rNode){
        if(lNode == NULL || rNode == NULL){
            return lNode == rNode;
        }
        if(lNode->val != rNode->val){
            return false;
        }
        return helper(lNode->left, rNode->right) && helper(lNode->right, rNode->left);
    }
};

/*
[]
[1]
[1,2,3]
[1,2,2,3,4,4,3]
[1,2,2,null,3,null,3]
*/
